csds-client/client/client.go

```diff
@@ -1,243 +1,22 @@
-			if err != nil {
-				return err
-			}
-
-			// parse GCP project number as header for authentication
-			var key string
-			switch c.info.uri {
-			case "trafficdirector.googleapis.com:443":
-				key = "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER"
-			}
-			if projectNum := getValueByKeyFromNodeMatcher(c.nodeMatcher, key); projectNum != "" {
-				c.metadata = metadata.Pairs("x-goog-user-project", projectNum)
-			}
-
-			c.clientConn, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
-			if err != nil {
-				return err
-			}
-			return nil
-		default:
-			return errors.New("auto authentication mode for this platform is not supported. Please use jwt_file instead")
-		}
-	default:
-		return errors.New("invalid authn_mode")
-	}
-}
-
-// New creates a new client
-func New() (*Client, error) {
-	c := &Client{
-		info: ParseFlags(),
-	}
-	if c.info.platform != "gcp" {
-		return nil, fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.info.platform)
-	}
-	if c.info.apiVersion != "v2" {
-		return nil, fmt.Errorf("%s api version is not supported, list of supported api versions: v2", c.info.apiVersion)
-	}
-
-	if err := c.parseNodeMatcher(); err != nil {
-		return nil, err
-	}
-
-	return c, nil
-}
-
-// Run connects the client to the uri and calls doRequest
-func (c *Client) Run() error {
-	if err := c.connWithAuth(); err != nil {
-		return err
-	}
-	defer c.clientConn.Close()
-
-	c.csdsClient = csdspb.NewClientStatusDiscoveryServiceClient(c.clientConn)
-	var ctx context.Context
-	if c.metadata != nil {
-		ctx = metadata.NewOutgoingContext(context.Background(), c.metadata)
-	} else {
-		ctx = context.Background()
-	}
-	streamClientStatus, err := c.csdsClient.StreamClientStatus(ctx)
-	if err != nil {
-		return err
-	}
-
-	// run once or run with monitor mode
-	for {
-		if err := c.doRequest(streamClientStatus); err != nil {
-			return err
-		}
-		if c.info.monitorInterval != 0 {
-			time.Sleep(c.info.monitorInterval)
-		} else {
-			if err := streamClientStatus.CloseSend(); err != nil {
-				return err
-			}
-			return nil
-		}
-	}
-}
-
-// doRequest sends request and print out the parsed response
-func (c *Client) doRequest(streamClientStatus csdspb.ClientStatusDiscoveryService_StreamClientStatusClient) error {
-
-	req := &csdspb.ClientStatusRequest{NodeMatchers: c.nodeMatcher}
-	if err := streamClientStatus.Send(req); err != nil {
-		return err
-	}
-
-	resp, err := streamClientStatus.Recv()
-	if err != nil && err != io.EOF {
-		return err
-	}
-
-	// post process response
-	if err := printOutResponse(resp,c.info); err != nil {
-		return err
-	}
-
-	return nil
+type ClientOptions struct {
+	Uri             string
+	Platform        string
+	AuthnMode       string
+	ApiVersion      string
+	RequestFile     string
+	RequestYaml     string
+	Jwt             string
+	ConfigFile      string
+	MonitorInterval time.Duration
+	Visualization   bool
+}
+
+type Client interface {
```

Comment:
Would be nice to add comments around it to describe what's expected from the interface.

Reply:
Added comments stated that `Packages which implement this interface in different api versions should have New() and Run() methods`

Reply:
Few comments about it:

New() is not dictated by the interface, it is package choice to provide a New() function.

Also comment should not repeat what's already clear from the code (in this case by looking at the interface I see that implementation must provide Run()), but rather add to the understanding. In this particular case, example of the comment could be:

For the interface:
// Client implements CSDS Client of a particular version. Upon creation of the new client it is expected that connection to the CSDS server is established.

For the run method:
// Run must send an CSDS request to the server and output the response according to the options provided during Client creation.

Reply:
Thanks for the example! Just added them to the code.